source/utilities.cc

```diff
@@ -1847,6 +1847,163 @@ namespace aspect
       return alpha;
     }
 
+    double
+    weighted_p_norm_average ( const std::vector<double> &weights,
+                              const std::vector<double> &values,
+                              const double p)
+    {
+      //TODO: prevent devision by zero for all
+      double averaged_parameter = 0.0;
+
+      // first look at the specal case which can be done faster
+      if (p <= -1000)
+        {
+          // Minimum
+          const unsigned int i = (unsigned int)(std::min_element( weights.begin(),
+                                                                  weights.end() )
+                                                - weights.begin());
+          return values[i];
+        }
+      else if (p == -1)
+        {
+          // Harmonic average (TODO: Check if we need to use a epsilon)
+          for (unsigned int i=0; i< weights.size(); ++i)
+            {
+              if (values[i] == 0 && weights[i] != 0)
+                return 0;
+              else if (values[i] !=0)
+                averaged_parameter += weights[i]/(values[i]);
+            }
+          const double sum_of_weights = std::accumulate(weights.rbegin(), weights.rend(), 0);
+          return sum_of_weights/averaged_parameter;
+        }
+      else if (p == 0)
+        {
+          // Geometric average (TODO: Check if we need to use a epsilon)
+          for (unsigned int i=0; i < weights.size(); ++i)
+            averaged_parameter += weights[i]*std::log(values[i]);
+          const double sum_of_weights = std::accumulate(weights.rbegin(), weights.rend(), 0);
+          return std::exp(averaged_parameter/sum_of_weights);
+        }
+      else if (p == 1)
+        {
+          // Arithmetic average (TODO: Check if we need to use a epsilon)
+          for (unsigned int i=0; i< weights.size(); ++i)
+            averaged_parameter += weights[i]*values[i];
```

Comment:
so this is not the l1 norm -- it's just the arithmetic average. there is nothing wrong with this, but it's worth not confusing this in the documentation by claiming that it's the l1 norm.